resources/2025-11-11-secrets-redactdata.md

```diff
@@ -0,0 +1,756 @@
+    User->>Frontend: PUT RRT with sensitive data<br/>{data: "secret123"}
+    Frontend->>Frontend: Validate request
+    Frontend->>Frontend: Encrypt fields marked<br/>x-radius-sensitive: true
+    Frontend->>DB: Save resource<br/>(with ENCRYPTED data)
+    activate DB
+    Note over DB: 🔒 Encrypted data in DB<br/>Exposure window starts
+    Frontend->>Queue: Queue async operation
+    Frontend-->>User: 202 Accepted<br/>provisioningState: Accepted
+
+    Note over User,SecretStore: Phase 2: Backend Processing (Asynchronous)
+    Queue->>Backend: Process operation
+    Backend->>DB: Read resource<br/>(with encrypted data)
+    DB-->>Backend: Return resource (encrypted)
+    Backend->>Backend: Decrypt properties.data<br/>IN MEMORY ONLY
+    Backend->>Backend: Nullify encrypted data:<br/>properties.data = null
+    Backend->>DB: Save resource<br/>(data: null)
+    deactivate DB
+    Note over DB: ✅ Exposure window ends<br/>Encrypted data removed
+    
+    Note over Backend: In-memory decrypted data<br/>used for recipe execution
+    Backend->>Backend: Extract recipe config<br/>(uses in-memory data)
+    Backend->>Recipe: Execute recipe<br/>(with in-memory decrypted data)
+    Recipe->>SecretStore: Deploy secrets<br/>(create KeyVault/K8s secret)
+    SecretStore-->>Recipe: Success + Output Resources
+    Recipe-->>Backend: Recipe output
+    
+    Note over Backend: Recipe complete
+    Backend->>Backend: Update status with<br/>outputResources
+    activate DB
+    Backend->>DB: Save resource<br/>(data: null, outputResources: [...])
+    deactivate DB
+    Backend->>Backend: Return Result
+    Note over Backend: On failure: Result{Requeue: false}<br/>On success: normal success result
+    Backend-->>User: Operation complete<br/>provisioningState: Succeeded
+
+    Note over User,SecretStore: Phase 3: Consumption
+    User->>SecretStore: Access secrets via<br/>outputResources reference
+    SecretStore-->>User: Secret values
+```
+
+### Detailed Design
+
+#### Option 1: Frontend Nullification (Rejected)
+
+**Description**: Nullify sensitive data in the frontend controller before initial database save, similar to Applications.Core/secretStores.
+
+**Reasons for rejection**:
+- **Cannot support recipes**: Recipe execution happens in backend async operation and needs access to sensitive data, this would break the recipe-based deployment.
+
+#### Option 2: Application-Layer Encryption with Backend Pre-Recipe Redaction  (Recommended)
+
+**Description**: The frontend encrypts sensitive fields marked with `x-radius-sensitive` using application-layer encryption before database storage. The backend controller loads the encrypted resource, decrypts sensitive fields into memory, immediately redacts the persisted record (saving `data: null` back to the database), and then runs the recipe using only the in-memory decrypted copy. The decrypted payload never leaves process memory. On failure, the controller returns `Result{Requeue=false}` to prevent retries.
```

Comment:
> The backend controller loads the encrypted resource

To clarify - the entire resource is not encrypted, only the fields marked as sensitive, is that correct?